cirq/circuits/circuit_operation.py

```diff
@@ -0,0 +1,407 @@
+            )
+        return new_op
+
+    def with_qubits(self, *new_qubits: 'cirq.Qid'):
+        """Returns a copy of this operation with an updated qubit mapping.
+
+        Args:
+            new_qubits: A list of qubits to target. Qubits in this list are
+                matched to qubits in the circuit following default qubit order,
+                ignoring any existing qubit map.
+
+        Returns:
+            A copy of this operation targeting `new_qubits`.
+
+        Raises:
+            ValueError: `new_qubits` has a different number of qubits than
+                this operation.
+        """
+        expected = protocols.num_qubits(self.circuit)
+        if len(new_qubits) != expected:
+            raise ValueError(f'Expected {expected} qubits, got {len(new_qubits)}.')
+        return self.with_qubit_mapping(dict(zip(self.qubits, new_qubits)))
+
+    def with_measurement_key_mapping(self, key_map: Dict[str, str]) -> 'CircuitOperation':
+        """Returns a copy of this operation with an updated key mapping.
+
+        Args:
+            key_map: A mapping of old measurement keys to new measurement keys.
+                This map will be composed with any existing key mapping.
+
+        Returns:
+            A copy of this operation with measurement keys updated as specified
+                by key_map.
+
+        Raises:
+            ValueError: The new operation has a different number of measurement
+                keys than this operation.
+        """
+        base_map = {m: m for m in self.circuit.all_measurement_keys()}
+        base_map.update(self.measurement_key_map)
+        new_map = {k: key_map.get(v, v) for k, v in base_map.items() if k != key_map.get(v, v)}
+        new_op = self.updated_copy(measurement_key_map=new_map)
+        if len(new_op._measurement_keys_()) != len(self._measurement_keys_()):
+            raise ValueError(
+                f'Collision in measurement key map composition. Original map:\n'
+                f'{self.measurement_key_map}\nApplied changes: {key_map}'
+            )
+        return new_op
+
+    def _with_measurement_key_mapping_(self, key_map: Dict[str, str]) -> 'CircuitOperation':
+        return self.with_measurement_key_mapping(key_map)
+
+    def with_params(
+        self, param_values: study.ParamResolverOrSimilarType, recursive: bool = False
+    ) -> 'CircuitOperation':
+        """Returns a copy of this operation with an updated ParamResolver.
+
+        Args:
+            param_values: A map or ParamResolver able to convert old param
+                values to new param values. This map will be composed with any
+                existing ParamResolver.
+
+        Returns:
+            A copy of this operation with its ParamResolver updated as specified
+                by param_values.
+        """
+        new_resolver = protocols.resolve_parameters(self.param_resolver, param_values, recursive)
+        return self.updated_copy(param_resolver=new_resolver)
```

Comment:
The way parameter resolution works on resolvers, we get some behavior here that is a bit odd. For example, suppose I have a circuit with parameters `a` and `b` and a map `a -> A` and `b -> B`, and I want to now map `A -> C`. I would expect to end up with `a -> C` and `b -> B` but the actual param resolver has an extra entry:
```
In [4]: cirq.resolve_parameters(cirq.ParamResolver({'a': 'A', 'b': 'B'}), {'A': 'C'})
Out[4]: cirq.ParamResolver({'A': sympy.Symbol('C'), 'a': sympy.Symbol('C'), 'b': sympy.Symbol('B')})
```
That extra `A -> C` entry is not needed here because `A` is not a parameter of the circuit. Similarly, if we compose with a map `A -> a` we get this extra entry and the identity remains in the map:
```
In [5]: cirq.resolve_parameters(cirq.ParamResolver({'a': 'A', 'b': 'B'}), {'A': 'a'})
Out[5]: cirq.ParamResolver({'A': sympy.Symbol('a'), 'a': sympy.Symbol('a'), 'b': sympy.Symbol('B')})
```
We could modify `ParamResolver._resolve_parameters_` to remove these identity elements, but in general param resolvers can't know the "full" key set they should apply to so removing the unneeded `A ->` entries that don't exist in the circuit is something we just need to do here, e.g.:
```python
new_params = {}
for k in protocols.parameter_symbols(self.circuit):
    v = self.param_resolver.value_of(v, recursive)
    v = protocols.resolve_parameters(v, param_values, recursive)
    if v != k:
        new_params[v] = k
return self.updated_copy(param_resolver=cirq.ParamResolver(new_params))
```

Reply:
The code snippet above will behave incorrectly for `recursive=True`:
```python
a, b = cirq.LineQubit.range(2)
x, y, z, X, Y, Z = [sympy.Symbol(x) for x in 'xyzXYZ']
op = cirq.CircuitOperation(cirq.CX(a, b) ** x, cirq.CZ(a, b) ** z)
op = op.with_params({x: X, z: Z}, recursive=True)
op2 = op.with_params({X: Y}, recursive=True)
op3 = op2.with_params({Z: X}, recursive=True)

# Removing mappings whose keys are not in the circuit:
assert op2.param_resolver.value_of(X) == X
assert op3.param_resolver.value_of(Z) == X

# Using ParamResolver composition:
assert op2.param_resolver.value_of(X) == Y
# {x: Y, X: Y, z: X, Z: X} flattens to {x: Y, X: Y, z: Y, Z: Y} when applied recursively
assert op3.param_resolver.value_of(Z) == Y
```
I'm okay with taking this as a sign that we should not support recursive parameter resolution in `CircuitOperation` (I'm frankly not fond of recursive resolution in general) but we'll need to be explicit about that.

Reply:
Single-step resolution is now enforced, with irrelevant parameters discarded.

Reply:
I've added a TODO here for graceful handling of recursive resolution, and opened #3619 to track it.